src/bucklescript/output_bucklescript_types.re

```diff
@@ -0,0 +1,419 @@
+  | Type(Object(_)) =>
+    raise(
+      Location.Error(
+        Location.error(~loc=loc |> conv_loc, "Object not allowed in args"),
+      ),
+    )
+  | Type(Union(_)) =>
+    raise(
+      Location.Error(
+        Location.error(~loc=loc |> conv_loc, "Union not allowed in args"),
+      ),
+    )
+  | Type(Interface(_)) =>
+    raise(
+      Location.Error(
+        Location.error(~loc=loc |> conv_loc, "Interface not allowed in args"),
+      ),
+    )
+  | Nullable(inner) =>
+    base_type(~inner=[generate_arg_type(loc, inner)], "option")
+  | List(inner) =>
+    base_type(~inner=[generate_arg_type(loc, inner)], "array")
+  | TypeNotFound(name) =>
+    raise(
+      Location.Error(
+        Location.error(
+          ~loc=loc |> conv_loc,
+          "Type " ++ name ++ " not found!",
+        ),
+      ),
+    );
+
+let generate_record_input_object = (input_obj_name, fields) => {
+  Ast_helper.Type.mk(
+    ~kind=
+      Ptype_record(
+        fields
+        |> List.map(
+             fun
+             | InputField({name, type_, loc}) => {
+                 Ast_helper.Type.field(
+                   {Location.txt: name, loc: Location.none},
+                   generate_arg_type(loc, type_),
+                 );
+               },
+           ),
+      ),
+    {
+      loc: Location.none,
+      txt:
+        generate_type_name(
+          ~prefix="t_variables",
+          switch (input_obj_name) {
+          | None => []
+          | Some(name) => [name]
+          },
+        ),
+    },
+  );
+};
+
+let generate_object_input_object = (input_obj_name, fields) => {
+  Ast_helper.(
+    Type.mk(
+      ~kind=Ptype_abstract,
+      ~manifest=
+        Typ.constr(
+          {Location.txt: Longident.parse("Js.t"), loc: Location.none},
+          [
+            Ast_helper.Typ.object_(
+              fields
+              |> List.map(
+                   fun
+                   | InputField({name, type_, loc}) =>
+                     Otag(
+                       {txt: name, loc: Location.none},
+                       [],
+                       generate_arg_type(loc, type_),
+                     ),
+                 ),
+              Closed,
+            ),
+          ],
+        ),
+      {
+        loc: Location.none,
+        txt:
+          generate_type_name(
+            ~prefix="t_variables",
+            switch (input_obj_name) {
+            | None => []
+            | Some(name) => [name]
+            },
+          ),
+      },
+    )
+  );
+};
+
+let generate_input_object =
+    (config: Generator_utils.output_config, input_obj_name, fields) => {
+  config.records
+    ? generate_record_input_object(input_obj_name, fields)
+    : generate_object_input_object(input_obj_name, fields);
+};
+
+let generate_arg_types = (config, variable_defs) => {
+  let input_objects = extract_args(config, variable_defs);
+
+  input_objects
+  |> List.map(
+       fun
+       | InputObject({name, fields}) => {
```

Comment:
No need for `fun` here. Normal destructuring should work.

Reply:
Actually here it needs to be, because it's a variant type with only one variant.